File: Benchmarks/Core/Systems/FontSystem2D.cs
using Benchmarks.Core.Components;
using Benchmarks.Core.Entity;
using Benchmarks.Core.Scene;
using Benchmarks.Core.Systems.Contract;
using Benchmarks.Core.Types;

namespace Benchmarks.Core.Systems
{
    /// <summary>
    /// Sistema encargado de dibujar texto en pantalla.
    /// Stub que replica el patron de iteracion del engine.
    /// </summary>
    public class FontSystem2D : ISystem
    {
        public int DrawCallCount { get; private set; }

        public override void InitializeSystem(EntityManager @entities)
        {
            EntityManager = @entities;
        }

        public override void SharedEntityInitialize(ref Yotsuba Entidad)
        {
            if (!Entidad.HasComponent(YTBComponent.Font)) return;
            ref FontComponent2D font = ref GetFontComponent(Entidad.Id);
            _ = font.IsVisible;
        }

        public void DrawSystem(GameTime gameTime)
        {
            DrawCallCount = 0;
            if (EntityManager == null) return;

            Span<FontComponent2D> fontComponents = GetFontComponentsAsSpan();
            Span<TransformComponent> transforms = GetTransformComponentsAsSpan();

            foreach (ref Yotsuba entity in GetEntitiesAsSpan())
            {
                if (!entity.HasComponent(YTBComponent.Font)) continue;

                ref readonly FontComponent2D font = ref fontComponents[entity.Id];
                if (!font.IsVisible) continue;

                ref readonly TransformComponent transform = ref transforms[entity.Id];

                // Simulate text draw
                _ = transform.Position;
                _ = font.Text;
                DrawCallCount++;
            }
        }

        public override void Dispose() { }
    }
}

File: Benchmarks_FSharp/Core/Systems/FontSystem2D.fs
namespace Benchmarks.Core.Systems

open Benchmarks.Core.Components
open Benchmarks.Core.Entity
open Benchmarks.Core.Scene
open Benchmarks.Core.Systems.Contract
open Benchmarks.Core.Types

type FontSystem2D() =
    inherit ISystem()

    member val DrawCallCount = 0 with get, set

    override this.InitializeSystem(entities: EntityManager) =
        this.EntityManager <- entities

    override this.SharedEntityInitialize(entidad: byref<Yotsuba>) =
        if not (entidad.HasComponent(YTBComponent.Font)) then ()
        else
        let font = &this.GetFontComponent(entidad.Id)
        font.IsVisible |> ignore

    member this.DrawSystem(gameTime: GameTime) =
        this.DrawCallCount <- 0
        if isNull this.EntityManager then ()
        else
        let fontComponents = this.GetFontComponentsAsSpan()
        let transforms = this.GetTransformComponentsAsSpan()
        let entities = this.GetEntitiesAsSpan()
        for i = 0 to entities.Length - 1 do
            let entity = &entities.[i]
            if entity.HasComponent(YTBComponent.Font) then
                let font = &fontComponents.[entity.Id]
                if font.IsVisible then
                    let transform = &transforms.[entity.Id]
                    transform.Position |> ignore
                    font.Text |> ignore
                    this.DrawCallCount <- this.DrawCallCount + 1

    override _.Dispose() = ()
